docs: gRPC tutorial (#3013)
bentoml/BentoML: docs/source/tutorial.rst

Tutorial: Intro to BentoML
==========================

> What are we building?

Reviewer: Can we incorporate the time commitment expected from users for the tutorial? I find it encouraging when documentation mention that the tutorial is expected to take 5, 10, or 15 minutes.

Reviewer: We should keep the tutorial in under 10 minutes.

---------------------

In this tutorial, we will focus on online model serving with BentoML, using a
classification model trained with `scikit-learn <https://scikit-learn.org/stable/>`_ and the Iris dataset.

> By the end of this tutorial, we will have an HTTP/gRPC endpoint for handling inference requests and a docker

Reviewer 2: ```suggestion
By the end of this tutorial, we will have a Bento that can be served easily using HTTP or gRPC for handling inference requests, and a docker
```

   If you have Docker installed, you can run the tutorial notebook from a pre-configured
   docker image with:

   .. code-block:: bash

>       » docker run -it --rm -p 8888:8888 -p 3000:3000 -p 3001:3001 bentoml/quickstart:latest

Reviewer 2: can't we use the docker expose ports thing since we use `EXPOSE`?

Author: EXPOSE is just documentation. We still need to specify the ports.

   .. code-block:: bash

      » pip install bentoml scikit-learn pandas

.. note::

>    BentoML provides gRPC support, in which we will demonstrate alongside with HTTP
>    implemenetation in this tutorial. However, this is optional and you don't have to
>    know about gRPC to get started with gRPC.
> 
>    If you are interested in walking through the gRPC example in this tutorial, install
>    the gRPC-variant of BentoML:

Reviewer 2: ```suggestion
   BentoML provides gRPC support, and we will provide gRPC examples alongside the HTTP
   ones in this tutorial. However, these examples are optional and you don't have to
   know about gRPC to get started with BentoML.

   If you are interested in trying the gRPC examples in this tutorial, install
   the gRPC dependencies for BentoML:
```

        return result

> Run the service in development mode:

Reviewer 2: ```suggestion
We can now run the BentoML server for our new service in development mode:
```

          2022-09-18T21:12:18-0700 [INFO] [cli] Prometheus metrics for gRPC BentoServer from "service.py:svc" can be accessed at http://0.0.0.0:3001.
          2022-09-18T21:12:18-0700 [INFO] [cli] Starting development gRPC BentoServer from "service.py:svc" running on http://0.0.0.0:3000 (Press CTRL+C to quit)
          2022-09-18 21:12:19 circus[81102] [INFO] Loading the plugin...
          2022-09-18 21:12:19 circus[81102] [INFO] Endpoint: 'tcp://127.0.0.1:61849'
          2022-09-18 21:12:19 circus[81102] [INFO] Pub/sub: 'tcp://127.0.0.1:61850'
          2022-09-18T21:12:19-0700 [INFO] [observer] BentoML is installed via development mode, adding source root to 'watch_dirs'
          2022-09-18T21:12:19-0700 [INFO] [observer] Watching directories: ['/Users/aarnphm/workspace/bentoml/aarnphm_fork/examples/quickstart', '/Users/aarnphm/bentoml/models', '/Users/aarnphm/workspace/bentoml/aarnphm_fork/bentoml']

.. dropdown:: About the command ``bentoml serve``
   :icon: code

   In the example above:

   .. tab-set::

      .. tab-item:: HTTP
         :sync: http

>          ``serve`` is the command to start a HTTP BentoServer.

Author: this is a dropdown, so verbose is fine?

>          ``serve`` is the command to start a HTTP BentoServer.

Reviewer: I'd say so. Users can read the CLI help text if needed. Our tutorial at the moment is a bit too complex and intimidating.

>          ``serve`` is the command to start a HTTP BentoServer.

Author: Got it

.. tab-set::

    .. tab-item:: HTTP
       :sync: http

       .. code-block:: bash

          » bentoml serve iris_classifier:latest --production

          2022-09-18T21:22:17-0700 [INFO] [cli] Environ for worker 0: set CPU thread count to 10
          2022-09-18T21:22:17-0700 [INFO] [cli] Prometheus metrics for HTTP BentoServer from "iris_classifier:latest" can be accessed at http://0.0.0.0:3000/metrics.
          2022-09-18T21:22:18-0700 [INFO] [cli] Starting production HTTP BentoServer from "iris_classifier:latest" running on http://0.0.0.0:3000 (Press CTRL+C to quit)

    .. tab-item:: gRPC
       :sync: grpc

       .. code-block:: bash

          » bentoml serve-grpc iris_classifier:latest --production

          2022-09-18T21:23:11-0700 [INFO] [cli] Environ for worker 0: set CPU thread count to 10
          2022-09-18T21:23:11-0700 [INFO] [cli] Prometheus metrics for gRPC BentoServer from "iris_classifier:latest" can be accessed at http://0.0.0.0:3001.
          2022-09-18T21:23:11-0700 [INFO] [cli] Starting production gRPC BentoServer from "iris_classifier:latest" running on http://0.0.0.0:3000 (Press CTRL+C to quit)

.. note::

>    Even though the service definition code uses model ``iris_clf:latest``, the
>    ``latest`` version can be resolved with local model store to find the exact model
>    version during the ``bentoml build`` process.
> 
>    This model is then bundled into the Bento, which makes sure this Bento is always using this exact model version, wherever it is deployed.

Reviewer: ```suggestion
   The build process resolves ``iris_clf:latest`` and packages the latest version of the ``iris_clf`` model in the model store to ensure the same version of the model gets deployed every time.
```
